plugins/cpp_metrics/service/src/cppmetricsservice.cpp

```diff
@@ -81,6 +96,111 @@ double CppMetricsServiceHandler::getSingleCppMetricForAstNode(
   });
 }
 
+void CppMetricsServiceHandler::getCppMetricsForModule(
+  std::vector<CppMetricsModule>& _return,
+  const core::FileId& fileId_)
+{
+  CppMetricsModule metric;
+
+  _transaction([&, this](){
+    typedef odb::query<model::CppFileMetrics> CppModuleMetricsQuery;
+
+    auto moduleMetrics = _db->query<model::CppFileMetrics>(
+      CppModuleMetricsQuery::file == std::stoull(fileId_));
+
+    for (const auto& moduleMetric : moduleMetrics)
+    {
+      metric.type = static_cast<CppModuleMetricsType::type>(moduleMetric.type);
+      metric.value = moduleMetric.value;
+      _return.push_back(metric);
+    }
+  });
+}
+
+void CppMetricsServiceHandler::getCppAstNodeMetricsForPath(
+  std::vector<CppAllMetricsAstNode>& _return,
+  const std::string& path_)
+{
+  _transaction([&, this]()
+  {
+    typedef odb::query<model::CppAstNodeMetrics> CppAstNodeMetricsQuery;
+    typedef odb::result<model::CppAstNodeMetrics> CppAstNodeMetricsResult;
+    typedef odb::query<model::CppAstNodeFilePath> CppAstNodeFilePathQuery;
+    typedef odb::result<model::CppAstNodeFilePath> CppAstNodeFilePathResult;
+
+    CppAstNodeFilePathResult nodes = _db->query<model::CppAstNodeFilePath>(
+      CppAstNodeFilePathQuery::LocFile::path.like(path_ + '%'));
+
+    if (nodes.empty())
+    {
+      core::InvalidInput ex;
+      ex.__set_msg("Invalid metric type for path: " + path_);
+      throw ex;
+    }
+
+    for (const auto& node : nodes)
+    {
+      auto metricsQuery = _db->query<model::CppAstNodeMetrics>(
+        CppAstNodeMetricsQuery::astNodeId == node.id);
+      std::vector<CppMetricsAstNode> metrics;
+
+      CppMetricsAstNode metricsAstNode;
+      for (const auto& metric : metricsQuery)
+      {
+        metricsAstNode.type = static_cast<CppAstNodeMetricsType::type>(metric.type);
+        metricsAstNode.value = metric.value;
+      }
+
+      CppAllMetricsAstNode nodeMetric;
+      nodeMetric.id = std::to_string(node.id);
+      nodeMetric.metrics = metrics;
+      _return.push_back(nodeMetric);
+    }
+  });
+}
+
+void CppMetricsServiceHandler::getCppFileMetricsForPath(
+  std::vector<CppAllMetricsModule>& _return,
+  const std::string& path_)
+{
+  _transaction([&, this]()
+  {
+    typedef odb::query<model::File> FileQuery;
+    typedef odb::result<model::File> FileResult;
+    typedef odb::query<model::CppFileMetrics> CppFileMetricsQuery;
+    typedef odb::result<model::CppFileMetrics> CppFileMetricsResult;
+
+    FileResult descendants = _db->query<model::File>(
+      FileQuery::path.like(path_ + '%'));
```

Comment:
Files with 0 zero metrics should not be returned in my opinion. If we put this restriction into the query, even performance can be boosted.

Reply:
I have added a quick temporal solution for this in 0936a195ad87688138a03977d8fcafa5ab62eef5.  
However, this does not modify the query, which would also add the expected performance boost, so it should be refactored.

Reply:
I have yet to figure out how to put a subquery into an ODB query, nevertheless, I transformed the query as you asked. I could only do it with another query - which creates some more technical debt. So "fixed" in https://github.com/Ericsson/CodeCompass/commit/eb36b1b2b60fbf6e7909cbba169bc16ba37719ff.

Reply:
I have performance concerns with this approach, as this will load all AST nodes into memory, which have a metric assigned. Even temporary merge into `master` (with a later fix) should be supported by a performance evaluation on e.g. CodeCompass itself.

The proper solution could be either a subquery or a JOIN statement. I guess a new ODB view will be required for that.